File: PowerQuery_Diff_Template/v1/automated/ts_xldiff_compare_months_changes/src/reader.ts
/**
 * Load the per-month tables out of an .xlsx workbook.
 *
 * Two layouts are supported, in this order of preference:
 *
 * 1. Real Excel Tables (ListObjects) named `tbl_Jan` ... `tbl_Dec` -- the layout the
 *    Power Query template expects, since `Excel.CurrentWorkbook()` only sees named tables.
 * 2. Plain worksheets titled `Jan` ... `Dec`, with a header row in row 1. This is the
 *    fallback for the far more common workbook that never had tables defined.
 *
 * Months that are absent are simply not returned; the diff engine treats a gap the same
 * way the M code does.
 */

import { existsSync, statSync } from "node:fs";
import { basename } from "node:path";

import {
  gridFromRange,
  gridFromUsedRange,
  readXlsx,
  type CellValue,
  type SheetData,
  type WorkbookData,
} from "./xlsx.js";

export type { CellValue };
export type Row = Record<string, CellValue>;

export const MONTHS: string[] = [
  "Jan", "Feb", "Mar", "Apr", "May", "Jun",
  "Jul", "Aug", "Sep", "Oct", "Nov", "Dec",
];

export const DEFAULT_TABLE_PREFIX = "tbl_";
export const DEFAULT_KEY = "ID";

/** The workbook cannot be diffed as-is (missing file, no months, missing key). */
export class WorkbookError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "WorkbookError";
  }
}

/** One month's rows, keyed by the diff key. */
export class MonthTable {
  constructor(
    public month: string,
    public columns: string[],
    public rows: Row[] = [],
    public source = "",
  ) {}

  /** Index the rows by key. Later duplicates win, matching a join's last-write. */
  byKey(key: string): Map<CellValue, Row> {
    const index = new Map<CellValue, Row>();
    for (const row of this.rows) {
      const value = row[key] ?? null;
      if (value !== null) {
        index.set(value, row);
      }
    }
    return index;
  }
}

/** Normalize a cell value: blank-ish strings become null, others are stripped. */
function clean(value: CellValue): CellValue {
  if (typeof value === "string") {
    const stripped = value.trim();
    return stripped === "" ? null : stripped;
  }
  return value;
}

/** Turn a rectangular block of cell values (header row first) into typed rows. */
function rowsFromGrid(grid: CellValue[][]): { columns: string[]; rows: Row[] } {
  if (grid.length === 0) {
    return { columns: [], rows: [] };
  }

  const headerCells = grid[0].map(clean);
  // Trailing unnamed columns are Excel padding, not data.
  while (headerCells.length > 0 && headerCells[headerCells.length - 1] === null) {
    headerCells.pop();
  }
  if (headerCells.length === 0) {
    return { columns: [], rows: [] };
  }

  const columns = headerCells.map((cell, i) => (cell !== null ? String(cell) : `Column${i + 1}`));

  const rows: Row[] = [];
  for (const raw of grid.slice(1)) {
    const values = raw.slice(0, columns.length).map(clean);
    if (values.every((value) => value === null)) {
      continue; // spacer row
    }
    while (values.length < columns.length) {
      values.push(null);
    }
    const row: Row = {};
    columns.forEach((column, i) => {
      row[column] = values[i];
    });
    rows.push(row);
  }

  return { columns, rows };
}

function readNamedTables(
  workbook: WorkbookData,
  prefix: string,
  months: readonly string[],
): Map<string, MonthTable> {
  const wanted = new Map<string, string>();
  for (const month of months) {
    wanted.set(`${prefix}${month}`.toLowerCase(), month);
  }

  const found = new Map<string, MonthTable>();
  for (const sheet of workbook.sheets) {
    for (const table of sheet.tables) {
      const month = wanted.get(table.name.toLowerCase());
      if (month === undefined) continue;

      const { columns, rows } = rowsFromGrid(gridFromRange(sheet, table.ref));
      if (columns.length > 0) {
        found.set(month, new MonthTable(month, columns, rows, `table ${table.name}`));
      }
    }
  }
  return found;
}

function readSheets(
  workbook: WorkbookData,
  months: readonly string[],
): Map<string, MonthTable> {
  const wanted = new Map<string, string>();
  for (const month of months) {
    wanted.set(month.toLowerCase(), month);
  }

  const found = new Map<string, MonthTable>();
  for (const sheet of workbook.sheets) {
    const month = wanted.get(sheet.name.trim().toLowerCase());
    if (month === undefined) continue;

    const { columns, rows } = rowsFromGrid(gridFromUsedRange(sheet));
    if (columns.length > 0) {
      found.set(month, new MonthTable(month, columns, rows, `sheet ${sheet.name}`));
    }
  }
  return found;
}

export interface ReadOptions {
  key?: string;
  tablePrefix?: string;
  months?: readonly string[];
}

/**
 * Read every month present in the workbook, keyed by month name.
 *
 * Throws WorkbookError if the file is unreadable, no month is found, or a month that
 * was found lacks the key column -- the diff is meaningless without a join key.
 */
export function readWorkbook(path: string, options: ReadOptions = {}): Map<string, MonthTable> {
  const key = options.key ?? DEFAULT_KEY;
  const tablePrefix = options.tablePrefix ?? DEFAULT_TABLE_PREFIX;
  const months = options.months ? [...options.months] : MONTHS;

  if (!existsSync(path) || !statSync(path).isFile()) {
    throw new WorkbookError(`No such workbook: ${path}`);
  }

  let workbook: WorkbookData;
  try {
    workbook = readXlsx(path);
  } catch (exc) {
    const cause = exc instanceof Error ? exc.message : String(exc);
    throw new WorkbookError(`Could not open ${basename(path)} as an .xlsx workbook: ${cause}`);
  }

  let tables = readNamedTables(workbook, tablePrefix, months);
  if (tables.size === 0) {
    tables = readSheets(workbook, months);
  }

  if (tables.size === 0) {
    throw new WorkbookError(
      `${basename(path)} has no month tables. Expected Excel tables named ` +
        `'${tablePrefix}Jan'...'${tablePrefix}Dec', or sheets named 'Jan'...'Dec'.`,
    );
  }

  const monthIndex = (month: string): number => months.indexOf(month);
  const missingKey = [...tables.values()]
    .filter((table) => !table.columns.includes(key))
    .map((table) => table.month)
    .sort((a, b) => monthIndex(a) - monthIndex(b));
  if (missingKey.length > 0) {
    throw new WorkbookError(
      `Key column '${key}' is missing from: ${missingKey.join(", ")}. ` +
        `Pass --key to name the column that identifies a row.`,
    );
  }

  const ordered = new Map<string, MonthTable>();
  for (const month of months) {
    const table = tables.get(month);
    if (table !== undefined) {
      ordered.set(month, table);
    }
  }
  return ordered;
}

export type { SheetData };

File: PowerQuery_Diff_Template/v1/automated/py_xldiff_compare_months_changes/src/py_xldiff_compare_months_changes/reader.py
"""Load the per-month tables out of an .xlsx workbook.

Two layouts are supported, in this order of preference:

1. Real Excel Tables (ListObjects) named ``tbl_Jan`` ... ``tbl_Dec`` -- the layout the
   Power Query template expects, since ``Excel.CurrentWorkbook()`` only sees named tables.
2. Plain worksheets titled ``Jan`` ... ``Dec``, with a header row in row 1. This is the
   fallback for the far more common workbook that never had tables defined.

Months that are absent are simply not returned; the diff engine treats a gap the same
way the M code does.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Dict, List, Optional, Sequence

from openpyxl import load_workbook
from openpyxl.utils.cell import range_boundaries

MONTHS: List[str] = [
    "Jan", "Feb", "Mar", "Apr", "May", "Jun",
    "Jul", "Aug", "Sep", "Oct", "Nov", "Dec",
]

DEFAULT_TABLE_PREFIX = "tbl_"
DEFAULT_KEY = "ID"


class WorkbookError(Exception):
    """The workbook cannot be diffed as-is (missing file, no months, missing key)."""


@dataclass
class MonthTable:
    """One month's rows, keyed by the diff key."""

    month: str
    columns: List[str]
    rows: List[Dict[str, Any]] = field(default_factory=list)
    source: str = ""

    def by_key(self, key: str) -> "Dict[Any, Dict[str, Any]]":
        """Index the rows by key. Later duplicates win, matching a join's last-write."""
        return {row[key]: row for row in self.rows if row.get(key) is not None}


def _clean(value: Any) -> Any:
    """Normalize a cell value: blank-ish strings become None, others are stripped."""
    if isinstance(value, str):
        stripped = value.strip()
        return stripped or None
    return value


def _rows_from_grid(grid: Sequence[Sequence[Any]]) -> "tuple[List[str], List[Dict[str, Any]]]":
    """Turn a rectangular block of cell values (header row first) into typed rows."""
    if not grid:
        return [], []

    header_cells = [_clean(cell) for cell in grid[0]]
    # Trailing unnamed columns are Excel padding, not data.
    while header_cells and header_cells[-1] is None:
        header_cells.pop()
    if not header_cells:
        return [], []

    columns = [str(cell) if cell is not None else f"Column{i + 1}" for i, cell in enumerate(header_cells)]

    rows: List[Dict[str, Any]] = []
    for raw in grid[1:]:
        values = [_clean(cell) for cell in raw[: len(columns)]]
        if all(value is None for value in values):
            continue  # spacer row
        values += [None] * (len(columns) - len(values))
        rows.append(dict(zip(columns, values)))

    return columns, rows


def _read_named_tables(workbook, prefix: str, months: Sequence[str]) -> Dict[str, MonthTable]:
    wanted = {f"{prefix}{month}".casefold(): month for month in months}
    found: Dict[str, MonthTable] = {}

    for worksheet in workbook.worksheets:
        # ws.tables is a TableList: .items() yields (name, ref) pairs, .values() the tables.
        for table in worksheet.tables.values():
            month = wanted.get(str(table.name).casefold())
            if month is None:
                continue

            min_col, min_row, max_col, max_row = range_boundaries(table.ref)
            grid = [
                [cell.value for cell in row]
                for row in worksheet.iter_rows(
                    min_row=min_row, max_row=max_row, min_col=min_col, max_col=max_col
                )
            ]
            columns, rows = _rows_from_grid(grid)
            if columns:
                found[month] = MonthTable(month, columns, rows, source=f"table {table.name}")

    return found


def _read_sheets(workbook, months: Sequence[str]) -> Dict[str, MonthTable]:
    wanted = {month.casefold(): month for month in months}
    found: Dict[str, MonthTable] = {}

    for worksheet in workbook.worksheets:
        month = wanted.get(str(worksheet.title).strip().casefold())
        if month is None:
            continue

        grid = [list(row) for row in worksheet.iter_rows(values_only=True)]
        columns, rows = _rows_from_grid(grid)
        if columns:
            found[month] = MonthTable(month, columns, rows, source=f"sheet {worksheet.title}")

    return found


def read_workbook(
    path: "str | Path",
    key: str = DEFAULT_KEY,
    table_prefix: str = DEFAULT_TABLE_PREFIX,
    months: Optional[Sequence[str]] = None,
) -> Dict[str, MonthTable]:
    """Read every month present in the workbook, keyed by month name.

    Raises WorkbookError if the file is unreadable, no month is found, or a month that
    was found lacks the key column -- the diff is meaningless without a join key.
    """
    path = Path(path)
    if not path.is_file():
        raise WorkbookError(f"No such workbook: {path}")

    months = list(months) if months else MONTHS

    try:
        workbook = load_workbook(path, data_only=True, read_only=False)
    except Exception as exc:  # openpyxl raises a grab-bag of errors on bad input
        raise WorkbookError(f"Could not open {path.name} as an .xlsx workbook: {exc}") from exc

    try:
        tables = _read_named_tables(workbook, table_prefix, months)
        if not tables:
            tables = _read_sheets(workbook, months)
    finally:
        workbook.close()

    if not tables:
        raise WorkbookError(
            f"{path.name} has no month tables. Expected Excel tables named "
            f"'{table_prefix}Jan'...'{table_prefix}Dec', or sheets named 'Jan'...'Dec'."
        )

    missing_key = sorted(
        (table.month for table in tables.values() if key not in table.columns),
        key=months.index,
    )
    if missing_key:
        raise WorkbookError(
            f"Key column {key!r} is missing from: {', '.join(missing_key)}. "
            f"Pass --key to name the column that identifies a row."
        )

    return {month: tables[month] for month in months if month in tables}
